lib/matplotlib/colors.py

```diff
@@ -1572,23 +1590,50 @@ def hillshade(self, elevation, vert_exag=1, dx=1, dy=1, fraction=1.):
             A 2d array of illumination values between 0-1, where 0 is
             completely in shadow and 1 is completely illuminated.
         """
-        # Azimuth is in degrees clockwise from North. Convert to radians
-        # counterclockwise from East (mathematical notation).
-        az = np.radians(90 - self.azdeg)
-        alt = np.radians(self.altdeg)
 
         # Because most image and raster GIS data has the first row in the array
         # as the "top" of the image, dy is implicitly negative.  This is
         # consistent to what `imshow` assumes, as well.
         dy = -dy
 
-        # Calculate the intensity from the illumination angle
-        dy, dx = np.gradient(vert_exag * elevation, dy, dx)
-        # The aspect is defined by the _downhill_ direction, thus the negative
-        aspect = np.arctan2(-dy, -dx)
-        slope = 0.5 * np.pi - np.arctan(np.hypot(dx, dy))
-        intensity = (np.sin(alt) * np.sin(slope) +
-                     np.cos(alt) * np.cos(slope) * np.cos(az - aspect))
+        # compute the normal vectors from the partial derivatives
+        e_dy, e_dx = np.gradient(vert_exag * elevation, dy, dx)
+
+        normal = np.empty(elevation.shape + (3,))
+        normal[...,0] = -e_dx
+        normal[...,1] = -e_dy
+        normal[...,2] = 1
+        normal /= _linalg_norm(normal, keepdims=True, axis=-1)
```

Comment:
All of this vectorized trig on `dx` and `dy` amounts to... concatenating and normalizing.

You can get here by tedious application of trig identities, or simply by geometric reasoning on how to obtain normal vectors.